nucypher/blockchain/eth/actors.py

```diff
@@ -1643,84 +1653,201 @@ def __init__(self,
 
+    def _get_max_bonus_bid_from_max_stake(self) -> int:
+        """Returns maximum allowed bid calculated from maximum allowed locked tokens"""
+        max_bonus_tokens = self.economics.maximum_allowed_locked - self.economics.minimum_allowed_locked
+        bonus_eth_supply = sum(self._all_bonus_bidders.values()) if self._all_bonus_bidders else self.worklock_agent.get_bonus_eth_supply()
+        bonus_worklock_supply = self.worklock_agent.get_bonus_lot_value()
+        max_bonus_bid = max_bonus_tokens * bonus_eth_supply // bonus_worklock_supply
+        return max_bonus_bid
+
+    def get_whales(self, force_read: bool = False) -> Dict[str, int]:
+        max_bonus_bid_from_max_stake = self._get_max_bonus_bid_from_max_stake()
+
+        bidders = dict()
+        for bidder, bid in self._get_all_bonus_bidders(force_read).items():
+            if bid > max_bonus_bid_from_max_stake:
+                bidders[bidder] = bid
+        return bidders
+
+    def _get_all_bonus_bidders(self, force_read: bool = False) -> dict:
+        if not force_read and self._all_bonus_bidders:
+            return self._all_bonus_bidders
+
+        bidders = self.worklock_agent.get_bidders()
+        min_bid = self.economics.worklock_min_allowed_bid
+
+        self._all_bonus_bidders = dict()
+        for bidder in bidders:
+            bid = self.worklock_agent.get_deposited_eth(bidder)
+            if bid > min_bid:
+                self._all_bonus_bidders[bidder] = bid - min_bid
+        return self._all_bonus_bidders
+
+    def _reduce_bids(self, whales: dict):
+
+        min_whale_bonus_bid = min(whales.values())
+        max_whale_bonus_bid = max(whales.values())
+
+        # first step - align at a minimum bid
+        if min_whale_bonus_bid != max_whale_bonus_bid:
+            whales = dict.fromkeys(whales.keys(), min_whale_bonus_bid)
+            self._all_bonus_bidders.update(whales)
+
+        bonus_eth_supply = sum(self._all_bonus_bidders.values())
+        bonus_worklock_supply = self.worklock_agent.get_bonus_lot_value()
+        max_bonus_tokens = self.economics.maximum_allowed_locked - self.economics.minimum_allowed_locked
+        if (min_whale_bonus_bid * bonus_worklock_supply) // bonus_eth_supply <= max_bonus_tokens:
+            raise self.BidderError(f"Internal error: At least one of bidders {whales} has allowable bid")
+
+        a = min_whale_bonus_bid * bonus_worklock_supply - max_bonus_tokens * bonus_eth_supply
+        b = bonus_worklock_supply - max_bonus_tokens * len(whales)
+        refund = -(-a//b)  # div ceil
+        min_whale_bonus_bid -= refund
+        whales = dict.fromkeys(whales.keys(), min_whale_bonus_bid)
+        self._all_bonus_bidders.update(whales)
+
+        return whales
+
+    def force_refund(self) -> dict:
+        self._ensure_cancellation_window(ensure_closed=True)
+
+        whales = self.get_whales()
+        if not whales:
+            raise self.BidderError(f"All bids are allowable, force refund is not needed")
```

Comment:
```suggestion
            raise self.BidderError(f"Force refund aborted: No whales detected and all bids qualify for claims.")
```
Better yet, an exception `class WhaleError` can be implemented several times here for a somewhat more clear  indication of the context.  :whale:

Reply:
✔️